File: experiments-local/experiment-mu-kimi3-brazil/runs/agent=opencode_language=typescript_model=openrouter/moonshotai/kimi-k3_tooling=none/rep1-failed/src/normalize.ts
/**
 * Text normalization utilities.
 *
 * The six source CSVs use conflicting conventions for team names and dates:
 *   - "Palmeiras-SP" (state suffix, no spaces)
 *   - "América - MG" (state suffix, spaced dash)
 *   - "America MG" / "Atletico Mineiro" / "EC Bahia" (BR-Football style)
 *   - "Nacional (URU)" / "Barcelona-EQU" (country tags)
 *   - "Sport Club Corinthians Paulista" (full legal name)
 * plus three date formats: ISO, ISO-with-time, and DD/MM/YYYY.
 *
 * Because the same real-world club (and match) appears in several files,
 * every raw name is mapped to a canonical club key via (a) a generic
 * surface normalization and (b) a curated alias table for Brazilian
 * league clubs whose surface forms diverge across files. Canonical keys
 * drive grouping, joining and cross-file deduplication; display names
 * keep their original UTF-8 characters for output.
 */

/** Strip diacritics: "São Paulo" -> "Sao Paulo", "Grêmio" -> "Gremio". */
export function foldAccents(s: string): string {
  return s.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
}

/**
 * Surface key: fold accents, lowercase, drop parenthetical tags and
 * legal-form words/tokens, collapse to alphanumerics. Trailing state /
 * country codes ("-SP", " - MG", "-EQU") are KEPT in the key at this
 * stage so that "Atlético-MG" and "Atlético-GO" do not collide; the
 * alias table below resolves them to stable club ids.
 */
function surfaceKey(raw: string): string {
  let s = foldAccents(raw).toLowerCase().trim();
  s = s.replace(/\([^)]*\)/g, " ");
  s = s.replace(
    /\b(esporte clube|sport club|futebol clube|clube de regatas|sociedade esportiva|foot[\s-]?ball club)\b/g,
    " ",
  );
  // Standalone short legal tokens: "EC Bahia", "Fortaleza FC", "Santa Cruz FC".
  s = s.replace(/\b(ec|fc|afc|cf)\b/g, " ");
  s = s.replace(/[^a-z0-9]+/g, "");
  return s;
}

/**
 * Curated alias table mapping surface keys to canonical club ids for
 * Brazilian clubs whose names diverge across the six source files.
 * Club ids are short human-readable slugs; unlisted surface forms fall
 * back to their surface key (state/country codes stripped, see below).
 */
const CLUB_ALIASES = new Map<string, string>([
  // --- Série A regulars ---
  ["palmeirassp", "palmeiras"], ["palmeiras", "palmeiras"],
  ["flamengorj", "flamengo"], ["flamengo", "flamengo"],
  ["fluminenserj", "fluminense"], ["fluminense", "fluminense"],
  ["corinthianssp", "corinthians"], ["corinthians", "corinthians"],
  ["corinthianspaulista", "corinthians"],
  ["saopaulosp", "saopaulo"], ["saopaulo", "saopaulo"],
  ["santossp", "santos"], ["santos", "santos"],
  ["gremiors", "gremio"], ["gremio", "gremio"],
  ["gremioportoalegrense", "gremio"],
  ["internacionalrs", "internacional"], ["internacional", "internacional"],
  ["cruzeiromg", "cruzeiro"], ["cruzeiro", "cruzeiro"],
  ["vascorj", "vasco"], ["vasco", "vasco"],
  ["vascodagamarj", "vasco"], ["vascodagama", "vasco"],
  ["botafogorj", "botafogo"], ["botafogo", "botafogo"],
  // --- Ambiguous "Atlético"/"America"/"Sport" family (state retained) ---
  ["atleticomg", "atletico-mg"], ["atleticomineiro", "atletico-mg"],
  ["atleticogo", "atletico-go"], ["atleticogoianiense", "atletico-go"],
  ["atleticopr", "athletico-pr"], ["athleticopr", "athletico-pr"],
  ["atleticoparanaense", "athletico-pr"], ["athleticoparanaense", "athletico-pr"],
  ["atleticoparanaensepr", "athletico-pr"], ["athleticoparanaensepr", "athletico-pr"],
  ["atleticomineiromg", "atletico-mg"], ["atleticogoianiensego", "atletico-go"],
  ["athletico", "athletico-pr"],
  ["americamg", "america-mg"], ["americaminasgerais", "america-mg"],
  ["americarn", "america-rn"],
  ["sportpe", "sport"], ["sport", "sport"], ["sportrecife", "sport"],
  ["sportdorecife", "sport"],
  // --- Other recurring national clubs ---
  ["bahiaba", "bahia"], ["bahia", "bahia"],
  ["vitoriaba", "vitoria"], ["vitoria", "vitoria"],
  ["fortalezace", "fortaleza"], ["fortaleza", "fortaleza"],
  ["cearace", "ceara"], ["ceara", "ceara"], ["cearasporting", "ceara"],
  ["chapecoensesc", "chapecoense"], ["chapecoense", "chapecoense"],
  ["goiasgo", "goias"], ["goias", "goias"],
  ["cuiabamt", "cuiaba"], ["cuiaba", "cuiaba"],
  ["coritibapr", "coritiba"], ["coritiba", "coritiba"],
  ["criciumasc", "criciuma"], ["criciuma", "criciuma"],
  ["figueirenscesc", "figueirense"], ["figueirense", "figueirense"],
  ["joinvillesc", "joinville"], ["joinville", "joinville"],
  ["avaisc", "avai"], ["avai", "avai"],
  ["pontepretasp", "pontepreta"], ["pontepreta", "pontepreta"],
  ["paranapr", "parana"], ["parana", "parana"],
  ["bragantinosp", "bragantino"], ["bragantino", "bragantino"],
  ["redbullbragantinosp", "bragantino"], ["redbullbragantino", "bragantino"],
  ["santacruzpe", "santacruz"], ["santacruz", "santacruz"],
  ["nauticope", "nautico"], ["nautico", "nautico"],
  ["guaranisp", "guarani"], ["guarani", "guarani"],
  ["portuguesasp", "portuguesa"], ["portuguesa", "portuguesa"],
  ["saocaetanosp", "saocaetano"], ["saocaetano", "saocaetano"],
  ["santoandresp", "santoandre"], ["santoandre", "santoandre"],
  ["juventuders", "juventude"], ["juventude", "juventude"],
  ["atleticoba", "atletico-ba"],
  ["gremiobaruerisp", "gremioprudente"], ["gremiobarueri", "gremioprudente"],
  ["gremioprudentesp", "gremioprudente"], ["gremioprudente", "gremioprudente"],
  ["ipatingamg", "ipatinga"], ["ipatinga", "ipatinga"],
  ["brasiliensedf", "brasiliense"], ["brasiliense", "brasiliense"],
  ["baruerisp", "barueri"], ["barueri", "barueri"],
  ["csaal", "csa"], ["csa", "csa"],
]);

/**
 * Canonical club key. Aliased clubs resolve to their club id; everything
 * else falls back to the surface key with trailing state/country codes
 * stripped (small state-league clubs, foreign clubs).
 */
export function canonicalTeamKey(raw: string): string {
  const surface = surfaceKey(raw);
  const aliased = CLUB_ALIASES.get(surface);
  if (aliased) return aliased;
  // Generic fallback: strip trailing "-XX" state/country tags.
  let s = foldAccents(raw).toLowerCase().trim();
  s = s.replace(/\([^)]*\)/g, " ");
  s = s.replace(
    /\b(esporte clube|sport club|futebol clube|clube de regatas|sociedade esportiva)\b/g,
    " ",
  );
  for (let i = 0; i < 3; i++) {
    s = s.replace(/\s*[-–—]\s*[a-z]{2,3}\s*$/i, "").trim();
  }
  s = s.replace(/\b(ec|fc|afc|cf)\b/g, " ");
  return s.replace(/[^a-z0-9]+/g, "");
}

/**
 * Produce a clean display name from a raw source name:
 * strips state/country suffixes and parenthetical tags but keeps accents.
 * "Palmeiras-SP" -> "Palmeiras", "Nacional (URU)" -> "Nacional".
 */
export function displayTeamName(raw: string): string {
  let s = raw.trim();
  s = s.replace(/\([^)]*\)/g, " ");
  for (let i = 0; i < 3; i++) {
    s = s.replace(/\s*[-–—]\s*[A-Z]{2,3}\s*$/, "").trim();
  }
  return s.replace(/\s+/g, " ").trim();
}

/**
 * Does `raw` refer to the team the user asked for (`query`)?
 * Matches on canonical key equality, or when one key contains the other
 * (so an unqualified "Remo" query still matches "Remo - PA").
 */
export function teamMatches(raw: string, query: string): boolean {
  const a = canonicalTeamKey(raw);
  const b = canonicalTeamKey(query);
  if (!a || !b) return false;
  return a === b || a.includes(b) || b.includes(a);
}

/**
 * Canonical display names for aliased clubs — unambiguous and correctly
 * accented ("Atlético-MG" vs "Atlético-GO", which would both display as
 * "Atlético" if the state suffix were simply stripped).
 */
const CLUB_DISPLAY = new Map<string, string>([
  ["palmeiras", "Palmeiras"], ["flamengo", "Flamengo"],
  ["fluminense", "Fluminense"], ["corinthians", "Corinthians"],
  ["saopaulo", "São Paulo"], ["santos", "Santos"],
  ["gremio", "Grêmio"], ["internacional", "Internacional"],
  ["cruzeiro", "Cruzeiro"], ["vasco", "Vasco"],
  ["botafogo", "Botafogo"], ["atletico-mg", "Atlético-MG"],
  ["atletico-go", "Atlético-GO"], ["athletico-pr", "Athletico-PR"],
  ["america-mg", "América-MG"], ["america-rn", "América-RN"],
  ["sport", "Sport"], ["bahia", "Bahia"], ["vitoria", "Vitória"],
  ["fortaleza", "Fortaleza"], ["ceara", "Ceará"],
  ["chapecoense", "Chapecoense"], ["goias", "Goiás"],
  ["cuiaba", "Cuiabá"], ["coritiba", "Coritiba"],
  ["criciuma", "Criciúma"], ["figueirense", "Figueirense"],
  ["joinville", "Joinville"], ["avai", "Avaí"],
  ["pontepreta", "Ponte Preta"], ["parana", "Paraná"],
  ["bragantino", "Bragantino"], ["santacruz", "Santa Cruz"],
  ["nautico", "Náutico"], ["guarani", "Guarani"],
  ["portuguesa", "Portuguesa"], ["saocaetano", "São Caetano"],
  ["santoandre", "Santo André"], ["juventude", "Juventude"],
  ["atletico-ba", "Atlético-BA"], ["gremioprudente", "Grêmio Prudente"],
  ["ipatinga", "Ipatinga"], ["brasiliense", "Brasiliense"],
  ["barueri", "Barueri"], ["csa", "CSA"],
]);

/**
 * Display name for a raw team name: canonical display for aliased clubs,
 * suffix-stripped original (accents kept) for everything else.
 */
export function canonicalDisplay(raw: string): string {
  const key = canonicalTeamKey(raw);
  return CLUB_DISPLAY.get(key) ?? displayTeamName(raw);
}

/**
 * Parse a date in any of the source formats into ISO YYYY-MM-DD.
 * Supported: "2023-09-24", "2012-05-19 18:30:00", "29/03/2003".
 * Returns null for unparseable input.
 */
export function parseDate(raw: string | null | undefined): string | null {
  if (!raw) return null;
  const s = raw.trim();
  if (!s) return null;

  // Brazilian DD/MM/YYYY
  const br = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})/);
  if (br) {
    const [, d, m, y] = br;
    return `${y}-${m.padStart(2, "0")}-${d.padStart(2, "0")}`;
  }

  // ISO date or ISO datetime: take the leading YYYY-MM-DD
  const iso = s.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (iso) return `${iso[1]}-${iso[2]}-${iso[3]}`;

  return null;
}

/** Extract the year from any supported date string. */
export function parseYear(raw: string | null | undefined): number | null {
  const iso = parseDate(raw);
  if (iso) return Number(iso.slice(0, 4));
  const s = (raw ?? "").trim();
  const y = s.match(/(\d{4})/);
  return y ? Number(y[1]) : null;
}

/** Lenient integer parse: handles "1", "1.0", "\"2\"" — null-safe. */
export function toInt(raw: unknown): number | null {
  if (raw === null || raw === undefined) return null;
  const n = Number(String(raw).trim());
  return Number.isFinite(n) ? Math.trunc(n) : null;
}

/** Lenient float parse (BR-Football-Dataset stores goals as "1.0"). */
export function toFloat(raw: unknown): number | null {
  if (raw === null || raw === undefined) return null;
  const n = Number(String(raw).trim());
  return Number.isFinite(n) ? n : null;
}

File: experiments-local/experiment-mu-glm53-provider/runs/agent=oc-parasail_language=python_model=openrouter/z-ai/glm-5.3_tooling=none/rep3/brazilian_soccer/normalize.py
"""Team-name normalization for Brazilian soccer datasets.

Why this module exists
----------------------
The six provided CSV files write the same club in many different ways:

* with a state suffix:          "Palmeiras-SP", "América - MG", "America MG"
* with the full legal name:     "Sport Club Corinthians Paulista"
* with a country qualifier:     "Nacional (URU)", "Barcelona-EQU", "Guaraní-PAR"
* with or without accents:      "São Paulo" vs "Sao Paulo", "Grêmio" vs "Gremio"
* with spelling drift:          "Athletico Paranaense" vs "Atlético Paranaense"

To answer questions across files, every raw name is folded to a single
canonical *key* (a lowercase, accent-stripped, alphanumeric-only string).

Resolution algorithm (``team_key``)
-----------------------------------
1. Compact the raw name *keeping* parentheses content and suffixes
   ("Guaraní (PAR)" -> "guaranipar") and look it up in ``ALIASES``.
2. Strip parenthetical content and any trailing 2-3 letter state/country
   code ("Palmeiras-SP" -> "Palmeiras") and compact again.
   * If the base names a canonical club (``CLUBS``) and the stripped state
     matches the club's home state (or there was no state), the canonical
     key is returned.
   * If the state differs, the club is a *namesake* (e.g. "Flamengo - PI"
     is not Rio's Flamengo) and the state is kept in the key
     ("flamengopi").
   * Unclaimed bases keep their state suffix too ("operariopr").
3. Anything unknown falls back to the compacted base name.

A ``TeamRegistry`` is built at load time from every name seen in the data;
it records the variants, source files and a human-friendly display name
for each key, and offers forgiving (substring / fuzzy) lookup for queries
like "palmeiras", "corinthians" or "sport club corinthians paulista".
"""

from __future__ import annotations

import re
import unicodedata
from collections import Counter
from dataclasses import dataclass, field
from difflib import SequenceMatcher
from typing import Iterable, Optional

__all__ = [
    "fold",
    "compact",
    "team_key",
    "team_state",
    "Club",
    "TeamInfo",
    "TeamRegistry",
    "build_registry",
    "CLUBS",
    "ALIASES",
]

# ---------------------------------------------------------------------------
# Text folding helpers
# ---------------------------------------------------------------------------


def fold(text: str) -> str:
    """Lowercase and strip accents ("São Paulo" -> "sao paulo")."""
    if not text:
        return ""
    decomposed = unicodedata.normalize("NFKD", text)
    stripped = "".join(ch for ch in decomposed if not unicodedata.combining(ch))
    return " ".join(stripped.lower().split())


def compact(text: str) -> str:
    """Fold and keep only alphanumerics ("Vasco da Gama - RJ" -> "vascodagamarj")."""
    return "".join(ch for ch in fold(text) if ch.isalnum())


_PAREN_RE = re.compile(r"\([^)]*\)")
# Trailing state/country code, e.g. "-SP", " - MG", " EQU", " - POA".
# The code must be upper-case in the original text so real words such as
# "Boa" or "Mixto" are never stripped.
_STATE_RE = re.compile(r"(?:[-–]|\s)\s*([A-Z]{2,3})$")


def _split_base(raw: str) -> tuple[str, Optional[str]]:
    """Remove parenthetical content and a trailing state/country code.

    Returns ``(base, state)`` where *state* is the stripped code (e.g.
    "SP", "MG", "EQU") or ``None``.
    """
    name = (raw or "").strip()
    name = _PAREN_RE.sub(" ", name).strip()
    match = _STATE_RE.search(name)
    if match:
        state = match.group(1)
        base = name[: match.start()].strip(" -–\t")
        if base:
            return base, state
    return name, None


# ---------------------------------------------------------------------------
# Canonical clubs
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Club:
    """A canonical club: display name plus its home state (Brazilian clubs)."""

    key: str
    display: str
    state: Optional[str] = None


CLUBS: dict[str, Club] = {
    club.key: club
    for club in [
        # Rio de Janeiro
        Club("flamengo", "Flamengo", "RJ"),
        Club("fluminense", "Fluminense", "RJ"),
        Club("botafogo", "Botafogo", "RJ"),
        Club("vascodagama", "Vasco da Gama", "RJ"),
        # São Paulo
        Club("corinthians", "Corinthians", "SP"),
        Club("palmeiras", "Palmeiras", "SP"),
        Club("saopaulo", "São Paulo", "SP"),
        Club("santos", "Santos", "SP"),
        Club("pontepreta", "Ponte Preta", "SP"),
        Club("portuguesa", "Portuguesa", "SP"),
        Club("guarani", "Guarani", "SP"),
        Club("saocaetano", "São Caetano", "SP"),
        Club("redbullbragantino", "Red Bull Bragantino", "SP"),
        Club("santoandre", "Santo André", "SP"),
        Club("barueri", "Barueri", "SP"),
        Club("gremionovorizontino", "Grêmio Novorizontino", "SP"),
        # Rio Grande do Sul
        Club("gremio", "Grêmio", "RS"),
        Club("internacional", "Internacional", "RS"),
        Club("juventude", "Juventude", "RS"),
        Club("criciuma", "Criciúma", "SC"),  # SC, see below
        # Paraná
        Club("coritiba", "Coritiba", "PR"),
        Club("atleticoparanaense", "Athletico Paranaense", "PR"),
        Club("parana", "Paraná", "PR"),
        # Minas Gerais
        Club("atleticomineiro", "Atlético Mineiro", "MG"),
        Club("cruzeiro", "Cruzeiro", "MG"),
        Club("americamineiro", "América Mineiro", "MG"),
        Club("ipatinga", "Ipatinga", "MG"),
        # Bahia
        Club("bahia", "Bahia", "BA"),
        Club("vitoria", "Vitória", "BA"),
        # Pernambuco
        Club("sportrecife", "Sport Recife", "PE"),
        Club("nautico", "Náutico", "PE"),
        Club("santacruz", "Santa Cruz", "PE"),
        # Ceará
        Club("ceara", "Ceará", "CE"),
        Club("fortaleza", "Fortaleza", "CE"),
        # Goiás
        Club("goias", "Goiás", "GO"),
        Club("atleticogoianiense", "Atlético Goianiense", "GO"),
        Club("vilanova", "Vila Nova", "GO"),
        # Alagoas
        Club("crb", "CRB", "AL"),
        Club("csa", "CSA", "AL"),
        # Pará
        Club("paysandu", "Paysandu", "PA"),
        Club("remo", "Remo", "PA"),
        # Santa Catarina
        Club("avai", "Avaí", "SC"),
        Club("chapecoense", "Chapecoense", "SC"),
        Club("figueirense", "Figueirense", "SC"),
        Club("joinville", "Joinville", "SC"),
        # Mato Grosso
        Club("cuiaba", "Cuiabá", "MT"),
        # Distrito Federal
        Club("brasiliense", "Brasiliense", "DF"),
        # Namesake clubs that must stay distinct from their famous cousins
        Club("americarn", "América-RN", "RN"),
        Club("botafogosp", "Botafogo-SP", "SP"),
        Club("botafogopb", "Botafogo-PB", "PB"),
    ]
}

# Aliases map the *compacted full raw name* (suffixes/parentheses included)
# to a canonical key.  They are only needed where the generic base-stripping
# rules cannot succeed: bare short names ("Vasco"), full legal names from
# the FIFA file, spelling drift ("Athletico" vs "Atlético"), state-abbreviated
# giants ("Atletico-MG") and Libertadores clubs carrying country codes.
ALIASES: dict[str, str] = {
    # Bare / short forms
    "vasco": "vascodagama",
    "sport": "sportrecife",
    "bragantino": "redbullbragantino",
    "bragantinosp": "redbullbragantino",
    # FIFA-style legal names
    "sportclubcorinthianspaulista": "corinthians",
    "sportclubdorecife": "sportrecife",
    "cearasportingclub": "ceara",
    "americafcminasgerais": "americamineiro",
    # BR-Football naming quirks
    "ecjuventude": "juventude",
    "clubedoremo": "remo",
    "nauticocapibaribe": "nautico",
    "fortalezaec": "fortaleza",
    "fortalezafc": "fortaleza",
    "santacruzfc": "santacruz",
    "vitoriaec": "vitoria",
    "novorizontino": "gremionovorizontino",
    "novorizontinosp": "gremionovorizontino",
    # Athletico Paranaense spellings (2019 rebrand kept both spellings)
    "athletico": "atleticoparanaense",
    "athleticopr": "atleticoparanaense",
    "athleticoparanaense": "atleticoparanaense",
    "athleticoparanaensepr": "atleticoparanaense",
    "atleticopr": "atleticoparanaense",
    "atleticoparanaensepr": "atleticoparanaense",
    # Atlético Mineiro / Goianiense abbreviated forms
    "atleticomg": "atleticomineiro",
    "atleticogo": "atleticogoianiense",
    # América Mineiro
    "americamg": "americamineiro",
    # Dotted acronym clubs (Copa do Brasil file)
    "crbal": "crb",
    "csaal": "csa",
    # Libertadores clubs written with country qualifiers; kept distinct
    # from any Brazilian namesake (e.g. Paraguay's Guaraní vs Guarani-SP).
    "nacionaluru": "nacionaluru",
    "nacionalpar": "nacionalpar",
    "guaranipar": "guaranipar",
    "barcelonaequ": "barcelonaequ",
    "delfinequ": "delfinequ",
    "riverplateuru": "riverplateuru",
    "olimpiapar": "olimpiapar",
    "libertadpar": "libertadpar",
    "universitarioper": "universitarioper",
    "trujillanosven": "trujillanosven",
}

# Display names for the self-mapped foreign clubs above (nice output).
_FOREIGN_DISPLAY = {
    "nacionaluru": "Nacional (URU)",
    "nacionalpar": "Nacional (PAR)",
    "guaranipar": "Guaraní (PAR)",
    "barcelonaequ": "Barcelona (EQU)",
    "delfinequ": "Delfín (EQU)",
    "riverplateuru": "River Plate (URU)",
    "olimpiapar": "Olimpia (PAR)",
    "libertadpar": "Libertad (PAR)",
    "universitarioper": "Universitario (PER)",
    "trujillanosven": "Trujillanos (VEN)",
}


# ---------------------------------------------------------------------------
# Key resolution
# ---------------------------------------------------------------------------


def team_key(raw: str) -> str:
    """Fold a raw team name to its canonical key."""
    name = (raw or "").strip()
    if not name:
        return ""
    full = compact(name)
    if full in ALIASES:
        return ALIASES[full]
    base, state = _split_base(name)
    base_key = compact(base)
    if base_key in CLUBS:
        club = CLUBS[base_key]
        if state is None or state.upper() == (club.state or ""):
            return club.key
        # Same nickname, different state: a namesake club.
        return base_key + state.lower()
    if base_key in ALIASES:
        canonical = ALIASES[base_key]
        club = CLUBS.get(canonical)
        if club and state and club.state and state.upper() != club.state:
            return base_key + state.lower()
        return canonical
    if state:
        return base_key + state.lower()
    return base_key or full


def team_state(raw: str) -> Optional[str]:
    """Extract the trailing state/country code from a raw name, if any."""
    _, state = _split_base((raw or "").strip())
    return state


# ---------------------------------------------------------------------------
# Registry
# ---------------------------------------------------------------------------


@dataclass
class TeamInfo:
    """Everything known about one resolved team key."""

    key: str
    display: str
    variants: list[str] = field(default_factory=list)
    states: set[str] = field(default_factory=set)
    sources: set[str] = field(default_factory=set)
    match_count: int = 0

    def add_variant(self, raw: str, source: str = "") -> None:
        if raw not in self.variants:
            self.variants.append(raw)
        if source:
            self.sources.add(source)
        state = team_state(raw)
        if state:
            self.states.add(state.upper())


def build_registry(raw_names: Iterable[tuple[str, str]]) -> TeamRegistry:
    """Build a registry from ``(raw_name, source_label)`` pairs."""
    registry = TeamRegistry()
    for raw, source in raw_names:
        if not raw or not raw.strip():
            continue
        key = team_key(raw)
        if not key:
            continue
        info = registry.teams.setdefault(key, TeamInfo(key=key, display=key))
        info.add_variant(raw.strip(), source)
    # Choose the nicest display name for every key.
    for key, info in registry.teams.items():
        if key in CLUBS:
            info.display = CLUBS[key].display
        elif key in _FOREIGN_DISPLAY:
            info.display = _FOREIGN_DISPLAY[key]
        else:
            info.display = _pick_display(info.variants)
    return registry


def _pick_display(variants: list[str]) -> str:
    """Pick the prettiest raw variant for an uncurated team."""

    def score(name: str) -> tuple[int, int, int]:
        base, _ = _split_base(name)
        base = base.strip() or name.strip()
        has_accent = any(unicodedata.combining(ch) for ch in unicodedata.normalize("NFKD", base))
        # prefer: accented forms, then shorter (suffix-stripped) names
        return (1 if has_accent else 0, -len(base), -len(name))

    return max(variants, key=score) if variants else ""


class TeamRegistry:
    """Lookup of canonical team keys with forgiving matching."""

    def __init__(self) -> None:
        self.teams: dict[str, TeamInfo] = {}

    def __contains__(self, key: str) -> bool:
        return key in self.teams

    def __len__(self) -> int:
        return len(self.teams)

    def get(self, key: str) -> Optional[TeamInfo]:
        return self.teams.get(key)

    def all(self) -> list[TeamInfo]:
        return sorted(self.teams.values(), key=lambda t: (-t.match_count, t.display))

    def find(self, query: str) -> list[TeamInfo]:
        """Return candidate teams for a user query, best matches first.

        Matching is exact-key first, then substring containment, then
        fuzzy similarity, so "palmeirass", "Sport Club Corinthians
        Paulista" and "atletico-mg" all resolve.
        """
        qc = compact(query)
        if not qc:
            return []
        exact: list[TeamInfo] = []
        starts: list[TeamInfo] = []
        contains: list[TeamInfo] = []
        fuzzy: list[tuple[float, TeamInfo]] = []
        for info in self.teams.values():
            key = info.key
            display_c = compact(info.display)
            variants_c = [compact(v) for v in info.variants]
            names = [key, display_c, *variants_c]
            if qc in (key, display_c) or qc in variants_c:
                exact.append(info)
                continue
            for n in names:
                if n.startswith(qc):
                    starts.append(info)
                    break
            else:
                for n in names:
                    if qc in n or n in qc:
                        contains.append(info)
                        break
                else:
                    best = max(
                        (SequenceMatcher(None, qc, n).ratio() for n in names),
                        default=0.0,
                    )
                    if best >= 0.75:
                        fuzzy.append((best, info))
        seen: set[str] = set()
        result: list[TeamInfo] = []
        for info in exact + starts + contains:
            if info.key not in seen:
                seen.add(info.key)
                result.append(info)
        for ratio, info in sorted(fuzzy, key=lambda p: -p[0]):
            if info.key not in seen:
                seen.add(info.key)
                result.append(info)
        # Prefer well-known teams when several candidates tie.
        result.sort(key=lambda i: (-i.match_count, i.display))
        return result[:12]

    def match_counts(self, counter: Counter) -> None:
        """Attach match counts (used to rank candidates)."""
        for key, count in counter.items():
            if key in self.teams:
                self.teams[key].match_count = count
